File: MATLAB/example2.m
%% Koopman Operator Tutorial -- Duffing Oscillator No. 2 Example
% 
% S. Servadio, D. Arnas, and R. Linares, “A Koopman Operator Tutorial with 
% Othogonal Polynomials.” arXiv, Jul. 14, 2022. doi: 10.48550/arXiv.2111.07485.

clear; close all; clc;
set(groot, 'defaulttextinterpreter','latex');  
set(groot, 'defaultAxesTickLabelInterpreter','latex');  
set(groot, 'defaultLegendInterpreter','latex');
addpath("src\")

%% Equation parameters
% Duffing Oscillator 
%
% x1dot = x2
% x2dot = -alpha*x1 - beta*x1^3 - delta*x2 + gamma*cos(omega*t)
%
% where x1 and x2 are the states and alpha, beta, delta are some
% parameters. gamma*cos(omega*t) = gamma*u(t) is the control in which gamma
% and omega are parameters. HOWEVER, for this example we neglect the
% controls and treat it as an autonomous system.

alpha = 1;
beta = 2.0;
delta = 0.3;
% gamma = 2.5;
% omega = 2;

%% Other settings
nt = 1000;

x10 = 1; % Initial q
x20 = 0; % Initial p
x0 = [x10 x20];

tf = 20; % Final time
tk = linspace(0,tf,nt);

%% Differential equation
% Elements in the array f(a,b,c):
% a: dimension where the derivative is performed
% b: term in the derivative
% c = 1: coefficient of the terms in xdot = f(x)
% c != 1: variable 'c' and its exponent in the polynomial term
Nx = 2;  % number of variables
Nt = 3;  % maximum number of terms in f(x)
fx = zeros(Nx,Nt,Nx+1);

% Coefficients
fx(:,:,1) = [
       1     0      0;
  -alpha -beta -delta
];
% Exponents of x1
fx(:,:,2) = [
    0 0 0;
    1 3 0
];
% Exponents of x2
fx(:,:,3) = [
    1 0 0;
    0 0 1;
];

%% Koopman Solution
c = 6;
KSol = KoopOpSol(c,fx,x0,tk);
x1 = KSol(1,:);
x2 = KSol(2,:);

%% ODE45 results for comparison
param.alpha = alpha;
param.beta = beta;
param.delta = delta;
[t,states] = ode45(@(t,x) duffing2noforce(t,x,param), tk, x0);


%% Plotting results
fig1 = figure(1);
scatter(x1,x2,DisplayName="Koopman")
hold on; 
scatter(states(:,1),states(:,2),"+",DisplayName="analytical")
legend;
hold off

fig2 = figure(2);
plot(tk,x1,lineWidth=2,DisplayName="$x_{1,k}$")
hold on
plot(tk,x2,lineWidth=2,DisplayName="$x_{2,k}$")
plot(tk,states(:,1),'--',lineWidth=2,DisplayName="$x_{1}$")
plot(tk,states(:,2),'--',lineWidth=2,DisplayName="$x_{2}$")
legend;
hold off


%% Function
function xdot = duffing2noforce(t,x,p)
    alpha = p.alpha; beta = p.beta; delta = p.delta;

    xdot = zeros(2,1);
    x1 = x(1);
    x2 = x(2);
    xdot(1) = x2;
    xdot(2) = -alpha*x1 - beta*x1^3 - delta*x2;
end

File: scripts/example2.jl
## Example 2: Duffing Oscillator (general form)

using DifferentialEquations
using KoopOpGalerkin
using GLMakie
# Makie.inline!(true)  # uncomment to show in navigation pane

const KOG = KoopOpGalerkin
const NORMALIZE = false
const REGULARIZE = false

## Equation parameters for the Duffing Oscillator
alpha = 1.0
beta = 2.0
delta = 0.4

# Other settings
nt = 1000
x10 = 1.0  # Initial x1
x20 = 0.0  # Initial x2
x0 = [x10, x20]
tf = 20.0  # Final time
tk = range(0, tf, length = nt)
dt = tk[2] - tk[1]

## Run ODE45 system to obtain the L∞ norm of x1 and x2
if NORMALIZE
    function duffing2noforce!(xdot, x, p, t)
        alpha, beta, delta = p
        x1 = x[1]
        x2 = x[2]
        xdot[1] = x2
        xdot[2] = -alpha * x1 - beta * x1^3 - delta * x2
        return xdot
    end
    param = [alpha, beta, delta]
    prob = ODEProblem(duffing2noforce!, x0, (0.0, tf), param)
    sol = solve(prob, Tsit5(), saveat = tk)
    x1max = maximum(abs.(sol[1, :]))
    x2max = maximum(abs.(sol[2, :]))
    x0 = x0 ./ [x1max, x2max]
else
    x1max = 1
    x2max = 1
end

## Differential equation
Nx = 2  # Number of variables
Nt = 3  # Maximum number of terms in f(x)
fx = zeros(Float64, Nx, Nt, Nx + 1)

# Coefficients and exponents
fx[:, :, 1] = [1.0 0.0 0.0; -alpha -beta -delta]
fx[:, :, 2] = [0.0 0.0 0.0; 1.0 3.0 0.0]
fx[:, :, 3] = [1.0 0.0 0.0; 0.0 0.0 1.0]

# Normalize
fx[1, :, 1] ./= x1max
fx[2, :, 1] ./= x2max

## Koopman Solution
c = 8
KSol = KOG.KoopmanSolve(c, fx, x0, tk)
x1 = KSol[1, :]
x2 = KSol[2, :]

## Normalized System: ODE45 results for comparison
function duffing2noforce!(xdot, x, p, t)
    alpha, beta, delta, x1max, x2max = p
    x1 = x[1]
    x2 = x[2]
    xdot[1] = (x2) / x1max
    xdot[2] = (-alpha * x1 - beta * x1^3 - delta * x2) / x2max
    return xdot
end
param = [alpha, beta, delta, x1max, x2max]
prob = ODEProblem(duffing2noforce!, x0, (0.0, tf), param)
sol = solve(prob, Tsit5(), saveat = tk)


## EDMD for comparison
# TODO: The EDMD results are incredibly off. Probably something wrong with the implementation.
# dsol = sol(tk, Val{1})  # generate derivative data
dsol = sol(range(tk.step.hi, tf+tk.step.hi, length=nt))

# Normalize the states and derivatives
lb_states = [minimum(sol[i,:]) for i = 1:Nx]
ub_states = [maximum(sol[i,:]) for i = 1:Nx]
states_n = KOG.normalize(sol[:,:], lb_states, ub_states, -1, 1)

lb_dstates = [minimum(dsol[i,:]) for i = 1:Nx]
ub_dstates = [maximum(dsol[i,:]) for i = 1:Nx]
dstates_n = KOG.normalize(dsol[:,:], lb_dstates, ub_dstates, -1, 1)

## Lift the states and derivatives
n_KO = 20
Xm_lift, Xp_lift, Ms = KOG.lift_data_RBF(states_n, dstates_n, n_KO; verbose=true)
if REGULARIZE
    Φ, Ω, Ã, ro = KOG.TREDMD(Xm_lift, Xp_lift, dt, 0.1, 5)
else
    Φ, Ω, Ã, ro = KOG.EDMD(Xm_lift, Xp_lift, dt, 3)
end

## Simulate the EDMD linearized system
x0_lift = KOG.lift_state_RBF(x0, n_KO, Ms)
sol_EDMD = zeros(ComplexF64, ro, nt)
b = Φ \ x0_lift
for i = 1:nt
    sol_EDMD[:, i] = exp.(Ω * tk[i]) .* b
end
sol_EDMD = Φ * sol_EDMD
sol_EDMD = real.(sol_EDMD)  # remove imaginary part

# Simulate the Ã matrix linear system from EDMD
# function duffing2noforce_EDMD!(dxdt, x, Ã, t)
#     dxdt .= Ã * x
# end
# x0_lift = KOG.lift_state_RBF(x0, n_KO, Ms)
# prob = ODEProblem(duffing2noforce_EDMD!, x0_lift[1:ro], (0.0, tf), Ã)
# sol_EDMD = solve(prob, Tsit5(), saveat = tk)

## Plotting results
fig = Figure(size=(800, 600))
ax1 = Axis(fig[1, 1], xlabel = "x1", ylabel="x2")
scatter!(ax1, sol[1, :], sol[2, :], color=:transparent, strokewidth=1, 
        strokecolor=:red, label="Analytical", marker=:circle)
scatter!(ax1, KSol[1, :], KSol[2, :], color=:blue, label="Koopman", marker=:cross)
scatter!(ax1, sol_EDMD[1, :], sol_EDMD[2, :], color=:transparent,
        strokecolor=:green, strokewidth=1, label="EDMD", marker=:diamond)
axislegend(ax1)
ax1.title = "Phase Space"

# Create line plot for time series
ax2 = Axis(fig[2, 1])
lines!(ax2, tk, sol[1, :], color = :red, label = "x1")
lines!(ax2, tk, sol[2, :], color = :purple, label = "x2")
lines!(ax2, tk, KSol[1, :], color = :blue, label = "x1_k", linestyle=:dash)
lines!(ax2, tk, KSol[2, :], color = :green, label = "x2_k", linestyle=:dash)
lines!(ax2, tk, sol_EDMD[1, :], color = :orange, label = "x1_EDMD", linestyle=:dashdot, linewidth=2)
lines!(ax2, tk, sol_EDMD[2, :], color = :pink, label = "x2_EDMD", linestyle=:dashdot, linewidth=2)
axislegend(ax2)
ax2.title = "Time Series"
ax2.xlabel = "t"
ax2.ylabel = "x"

# Display the figure
fig